src/Std/Data/Iterators/Lemmas/Consumers/Collect.lean

```diff
@@ -0,0 +1,111 @@
+    it.toListRev = it.toIterM.toListRev :=
+  rfl
+
+@[simp]
+theorem IterM.toList_toPureIter {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    {it : IterM (α := α) Id β} :
+    it.toPureIter.toList = it.toList :=
+  rfl
+
+@[simp]
+theorem IterM.toListRev_toPureIter {α β} [Iterator α Id β] [Finite α Id]
+    {it : IterM (α := α) Id β} :
+    it.toPureIter.toListRev = it.toListRev :=
+  rfl
+
+theorem Iter.toList_toArray {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toArray.toList = it.toList := by
+  simp only [toArray_eq_toArray_toIterM, toList_eq_toList_toIterM, ← IterM.toList_toArray,
+    Id.map_eq]
+
+theorem Iter.toArray_toList {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toList.toArray = it.toArray := by
+  simp only [toArray_eq_toArray_toIterM, toList_eq_toList_toIterM, ← IterM.toArray_toList,
+    Id.map_eq]
+
+theorem Iter.toListRev_eq {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toListRev = it.toList.reverse := by
+  simp [Iter.toListRev_eq_toListRev_toIterM, Iter.toList_eq_toList_toIterM, IterM.toListRev_eq]
+
+theorem Iter.toArray_of_step {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toArray = match it.step with
+      | .yield it' out _ => #[out] ++ it'.toArray
+      | .skip it' _ => it'.toArray
+      | .done _ => #[] := by
+  simp only [Iter.toArray_eq_toArray_toIterM, Iter.step]
+  rw [IterM.toArray_of_step]
+  simp only [Id.map_eq, Id.pure_eq, Id.bind_eq, Id.run]
+  generalize it.toIterM.step = step
+  cases step.casesHelper <;> simp
+
+theorem Iter.toList_of_step {α β} [Iterator α Id β] [Finite α Id] [IteratorCollect α Id]
+    [LawfulIteratorCollect α Id] {it : Iter (α := α) β} :
+    it.toList = match it.step with
+      | .yield it' out _ => out :: it'.toList
+      | .skip it' _ => it'.toList
+      | .done _ => [] := by
+  rw [← Iter.toList_toArray, Iter.toArray_of_step]
+  split <;> simp [Iter.toList_toArray]
+
+theorem Iter.toListRev_of_step {α β} [Iterator α Id β] [Finite α Id] {it : Iter (α := α) β} :
+    it.toListRev = match it.step with
+      | .yield it' out _ => it'.toListRev ++ [out]
+      | .skip it' _ => it'.toListRev
+      | .done _ => [] := by
+  rw [Iter.toListRev_eq_toListRev_toIterM, IterM.toListRev_of_step, Iter.step]
+  simp only [Id.map_eq, Id.pure_eq, Id.bind_eq, Id.run]
+  generalize it.toIterM.step = step
+  cases step.run.casesHelper <;> simp
+
+theorem Iter.getElem?_toList_eq_seekIdx? {α β}
+    [Iterator α Id β] [Finite α Id] [IteratorCollect α Id] [LawfulIteratorCollect α Id]
+    {it : Iter (α := α) β} {k : Nat} :
+    it.toList[k]? = it.seekIdx? k := by
+  revert k
+  induction it using Iter.induct with | step it ihy ihs =>
+  intro k
```

Comment:
Is this not just 
```suggestion
  induction it using Iter.induct generalizing k with | step it ihy ihs =>
```

Reply:
Neat. I tried `induction it generalizing k using Iter.induct` and gave up after that didn't work.